src/algebra/order/to_interval_mod.lean

```diff
@@ -457,6 +457,67 @@ begin
     rw [hz, to_Ioc_mod_zsmul_add] }
 end
 
+section Ico_Ioc
+
+variables (a : α) {b : α} (hb : 0 < b)
+
+lemma tfae_to_Ico_eq_to_Ioc (x : α) :
+  tfae
+   [to_Ico_div a hb x = to_Ioc_div a hb x,
+    to_Ico_div a hb x + 1 ≠ to_Ioc_div a hb x,
+    to_Ico_mod a hb x = to_Ioc_mod a hb x,
+    to_Ico_mod a hb x + b ≠ to_Ioc_mod a hb x,
+    to_Ioc_mod a hb x ≠ a + b,
+    to_Ico_mod a hb x ≠ a,
+    ∀ z : ℤ, a - x ≠ z • b,
+    ∃ z : ℤ, x + z • b ∈ set.Ioo a (a + b),
+    (x : α ⧸ add_subgroup.zmultiples b) ≠ a] :=
```

Comment:
This all feels pretty awkward to work with. Do you think it make sense to define a `mem_Ioo_mod` predicate, and expose each of these tfaes as a separate lemma?

Reply:
What's awkward about them exactly? I used this lemma in various proofs and feel good about it. Equivalences between inequalities would better be restated as equivalence between equalities, though.

Reply:
It's a lot harder to work out what `tfae.out X Y` means in a proof than a lemma with name `mem_Ioo_mod_iff_to_Ico_div_eq_to_Ioc_div` etc

Reply:
Okay, so your concern is proof readability. I'd need to extract four equivalences to replace all tfae.out. As a general solution, we should extend `alias` to work with tfae, perhaps? The number of total lemmas grow quadratically, but maybe we can use to_Ico_mod = to_Ioc_mod (2) as an intermediary and only stare its equivalence with the other conditions. (In my four uses, three involve (2) and the other is between (4) and (5).)

Reply:
> but maybe we can use to_Ico_mod = to_Ioc_mod (2) as an intermediary and restate only its equivalence with the other conditions

This was basically my suggestion; any TFAE with N entries can be restated as a new definition `X` and N-1 iff lemmas; with the advantage that it's easier to add more lemmas later without the TFAE growing and growing.
You could argue that this is exactly what we do with `odd` and `even`, which can be stated in lots of ways.

My claim is that `∃ z : ℤ, x + z • b ∈ set.Ioo a (a + b)` is the more elementary definition, but I don't feel strongly about that.

Reply:
Some equivalences in the tfae are proven using a cycle of implications though, and there isn't a easy way to directly get equivalence.

Reply:
And tfae readability issue is ticket #3532

Reply:
> Some equivalences in the tfae are proven using a cycle of implications though, and there isn't a easy way to directly get equivalence.

Because of this, I think I'll still first prove the TFAE and restate individual equivalences afterwards. I still think `to_Ico_mod = to_Ioc_mod` is a better central hub, so I attempted to state its equivalence with `mem_Ioo_mod` in [commit `1db30f2`](https://github.com/leanprover-community/mathlib/pull/17983/commits/1db30f2d931b5415138ba71d7e9c73e90226b90e). However, it's rather annoying that I have to write the full statement; Lean can't infer the type just from
```
include a hb
lemma to_Ico_mod_eq_to_Ioc_mod_iff_mem_Ioo_mod (x : α) := (tfae_to_Ico_eq_to_Ioc a hb x).out 2 7
```
Is there any meta code to make this work?

Reply:
Oh, actually changing `lemma` to `def` works. Is it possible to change the way Lean processes `lemma` declarations to make it able to infer the type for `lemma`, or is there a way to disable the [linter.def_lemma](https://leanprover-community.github.io/mathlib_docs/tactic/lint/misc.html#linter.def_lemma) in a section?

Reply:
I don't think you need most of the tfae here; once you're past the cyclic part of the argument, the rest is just as easy to write with regular lemmas.

> I still think to_Ico_mod = to_Ioc_mod is a better central hub,

In terms of being easiest to prove everything else from, or being the more natural definitionally? Only the latter really matters

Reply:
Okay, I agree that the `mem_Ioo_mod` predicate is probably best definitionally. However, the cycle currently involves 6 of the 9 conditions, and I'm still looking for a way to extract the 5 lemmas without typing the types.

Reply:
Okay, I've introduced the `mem_Ioo_mod` predicate, cut down the TFAE to four propositions, extracted three iff statements, proved the remaining five iffs as individual lemmas, and removed the use of tfae.out except in the three extracted iffs, in the last three commits. @eric-wieser What do you think?

Personally I still like tfae.out better, because it generally [makes proofs shorter](https://github.com/leanprover-community/mathlib/pull/17983/commits/0a60f2417fa05ec6ab5b4bfe17df908c25e173c9#diff-289dc9987f8c37ae4d6fc491bf58a0d8446444f6cedc0e819d27e46379dc8ae9R462-R467).

Reply:
Any other comments @eric-wieser ?

Reply:
I've resolved the merge conflict. If @eric-wieser has no other comments, maybe @sgouezel could approve?

Reply:
I'll take a look today or tomorrow

Reply:
@eric-wieser Any movement on this? This PR is blocking #17913 and #17994. Let me know if there's anything I can do to move things forward.